app/src/main/java/org/dhis2/data/forms/dataentry/fields/coordinate/CoordinateViewModel.java

```diff
@@ -88,4 +97,134 @@ public void onCurrentLocationClick(Geometry geometry) {
+
+    public void onLongitudeFocusChanged(boolean hasFocus, String currentLongitudeValue, String errorMessage, Function1<String, Unit> callback) {
+        if (!hasFocus) {
+            if (!currentLongitudeValue.isEmpty()) {
+                Double lonString = Double.valueOf(StringExtensionsKt.parseToDouble(currentLongitudeValue));
+                Double lon = DoubleExtensionsKt.truncate(lonString);
+                String error;
+                if (!LngLatValidatorKt.isLongitudeValid(lon)) {
+                    error = errorMessage;
+                } else {
+                    error = null;
+                }
+                callback.invoke(lon.toString());
+                observableErrorMessage.set(error);
+            }
+        } else {
+            onItemClick();
+        }
+    }
+
+    public boolean isPoint() {
+        return featureType() == FeatureType.POINT;
+    }
+
+    public void onDescriptionClick() {
+        callback.showDialog(label(), description());
+    }
+
+    public Geometry currentGeometry() {
+        if (featureType() == null)
+            throw new NullPointerException("use setFeatureType before setting an initial value");
+        if (value() != null) {
+            return Geometry.builder()
+                    .coordinates(value())
+                    .type(featureType())
+                    .build();
+        } else {
+            return null;
+        }
+    }
+
+    public boolean allowClearValue() {
+        return currentGeometry() != null && editable();
+    }
+
+    public void onClearValueClick() {
+        if (!activated()) {
+            onItemClick();
+        }
+        onCurrentLocationClick(null);
+    }
+
+    public boolean validateFilledCoordinates() {
+        return (!isEmpty(latitudeValue.get()) && !isEmpty(longitudeValue.get())) ||
+                (isEmpty(latitudeValue.get()) && isEmpty(longitudeValue.get()));
+    }
+
+    public void onLatitudeChanged(CharSequence s, int start, int before, int count) {
+        String latitude = s.toString();
+        latitudeValue.set(latitude.isEmpty() ? null : latitude);
+    }
+
+    public void onLongitudeChanged(CharSequence s, int start, int before, int count) {
+        String longitude = s.toString();
+        longitudeValue.set(longitude.isEmpty() ? null : longitude);
+    }
+
+    public boolean onFilledCoordinate(String coordinatesErrorMessage) {
+        if (validateFilledCoordinates()) {
+            Double lat = parseCoordinateStringToDouble(latitudeValue.get());
+            Double lon = parseCoordinateStringToDouble(longitudeValue.get());
+            if (lat != null && lon != null) {
+                if (LngLatValidatorKt.areLngLatCorrect(lat, lon)) {
+                    onCurrentLocationClick(GeometryHelper.createPointGeometry(lon, lat));
+                } else {
+                    observableErrorMessage.set(coordinatesErrorMessage);
+                }
+            } else {
+                onCurrentLocationClick(null);
```

Comment:
If lat and lon are null should we show observableErrorMessage.set(coordinatesErrorMessage) ??? Does it make sense to move  if (lat != null && lon != null) inside LngLatValidatorKt.areLngLatCorrect ????

Reply:
Nop, they can be both null for example if the user deletes them. An error only needs to be show if:
- Only one is provided
- Both are provided but does not pass the validation